File: 2025/day08/java/Solution.java
import java.io.*;
import java.util.*;

public class Solution {

    static class Point {
        int x, y, z;

        Point(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }
    }

    static class UnionFind {
        int[] parent;
        int[] rank;
        int[] size;

        UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            size = new int[n];

            for (int i = 0; i < n; i++) {
                parent[i] = i;
                rank[i] = 0;
                size[i] = 1;
            }
        }

        int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]); // Path compression
            }
            return parent[x];
        }

        boolean union(int x, int y) {
            int px = find(x);
            int py = find(y);

            if (px == py) {
                return false; // Already in same set
            }

            // Union by rank
            if (rank[px] < rank[py]) {
                int temp = px;
                px = py;
                py = temp;
            }

            parent[py] = px;
            size[px] += size[py];

            if (rank[px] == rank[py]) {
                rank[px]++;
            }

            return true;
        }

        List<Integer> getComponentSizes() {
            var sizes = new ArrayList<Integer>();
            for (int i = 0; i < parent.length; i++) {
                if (parent[i] == i) { // Root of a component
                    sizes.add(size[i]);
                }
            }
            return sizes;
        }
    }

    static class Edge implements Comparable<Edge> {
        long distSq;
        int i, j;

        Edge(long distSq, int i, int j) {
            this.distSq = distSq;
            this.i = i;
            this.j = j;
        }

        @Override
        public int compareTo(Edge other) {
            return Long.compare(this.distSq, other.distSq);
        }
    }

    static List<Point> parseInput(String filename) throws IOException {
        var points = new ArrayList<Point>();

        try (var br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (!line.isEmpty()) {
                    var parts = line.split(",");
                    int x = Integer.parseInt(parts[0]);
                    int y = Integer.parseInt(parts[1]);
                    int z = Integer.parseInt(parts[2]);
                    points.add(new Point(x, y, z));
                }
            }
        }

        return points;
    }

    static long euclideanDistanceSquared(Point p1, Point p2) {
        long dx = (long)p1.x - p2.x;
        long dy = (long)p1.y - p2.y;
        long dz = (long)p1.z - p2.z;
        return dx * dx + dy * dy + dz * dz;
    }

    static long part1(List<Point> points, int numConnections) {
        int n = points.size();

        // Generate all pairs with distances
        var edges = new ArrayList<Edge>();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                long distSq = euclideanDistanceSquared(points.get(i), points.get(j));
                edges.add(new Edge(distSq, i, j));
            }
        }

        // Sort by distance
        Collections.sort(edges);

        // Union-Find to connect closest pairs
        var uf = new UnionFind(n);
        int connections = 0;

        for (var edge : edges) {
            uf.union(edge.i, edge.j); // Always attempt union
            connections++;
            if (connections == numConnections) {
                break;
            }
        }

        // Get component sizes and find the 3 largest
        var sizes = uf.getComponentSizes();
        Collections.sort(sizes, Collections.reverseOrder());

        // Multiply the 3 largest
        return (long)sizes.get(0) * sizes.get(1) * sizes.get(2);
    }

    static long part2(List<Point> points) {
        int n = points.size();

        // Generate all pairs with distances
        var edges = new ArrayList<Edge>();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                long distSq = euclideanDistanceSquared(points.get(i), points.get(j));
                edges.add(new Edge(distSq, i, j));
            }
        }

        // Sort by distance
        Collections.sort(edges);

        // Union-Find to connect until all in one circuit
        var uf = new UnionFind(n);
        int numComponents = n;

        for (var edge : edges) {
            if (uf.union(edge.i, edge.j)) { // Actually merged two components
                numComponents--;
                if (numComponents == 1) {
                    // This was the last connection - all in one circuit now
                    var p1 = points.get(edge.i);
                    var p2 = points.get(edge.j);
                    return (long)p1.x * p2.x; // Product of X coordinates
                }
            }
        }

        return 0;
    }

    public static void main(String[] args) throws IOException {
        var inputFile = args.length > 0 ? args[0] : "../input.txt";
        var points = parseInput(inputFile);

        System.out.println("Part 1: " + part1(points, 1000));
        System.out.println("Part 2: " + part2(points));
    }
}

File: 2025/day08/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

function parseInput(filename) {
    const content = readFileSync(filename, 'utf-8');
    const points = [];
    for (const line of content.trim().split('\n')) {
        if (line.trim()) {
            const [x, y, z] = line.split(',').map(Number);
            points.push([x, y, z]);
        }
    }
    return points;
}

class UnionFind {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = new Array(n).fill(0);
        this.size = new Array(n).fill(1);
    }

    find(x) {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]); // Path compression
        }
        return this.parent[x];
    }

    union(x, y) {
        let px = this.find(x);
        let py = this.find(y);
        if (px === py) return false; // Already in same set

        // Union by rank
        if (this.rank[px] < this.rank[py]) {
            [px, py] = [py, px];
        }
        this.parent[py] = px;
        this.size[px] += this.size[py];
        if (this.rank[px] === this.rank[py]) {
            this.rank[px]++;
        }
        return true;
    }

    getComponentSizes() {
        const sizes = [];
        for (let i = 0; i < this.parent.length; i++) {
            if (this.parent[i] === i) {
                sizes.push(this.size[i]);
            }
        }
        return sizes;
    }
}

function distanceSq(p1, p2) {
    return (p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2 + (p1[2] - p2[2]) ** 2;
}

function part1(points, numConnections = 1000) {
    const n = points.length;

    // Generate all pairs with distances
    const pairs = [];
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            pairs.push([distanceSq(points[i], points[j]), i, j]);
        }
    }

    // Sort by distance
    pairs.sort((a, b) => a[0] - b[0]);

    // Union-Find to connect closest pairs
    const uf = new UnionFind(n);
    let connections = 0;
    for (const [distSq, i, j] of pairs) {
        uf.union(i, j); // Always attempt union
        connections++;
        if (connections === numConnections) break;
    }

    // Get component sizes and find 3 largest
    const sizes = uf.getComponentSizes().sort((a, b) => b - a);

    // Multiply the 3 largest
    return sizes[0] * sizes[1] * sizes[2];
}

function part2(points) {
    const n = points.length;

    // Generate all pairs with distances
    const pairs = [];
    for (let i = 0; i < n; i++) {
        for (let j = i + 1; j < n; j++) {
            pairs.push([distanceSq(points[i], points[j]), i, j]);
        }
    }

    // Sort by distance
    pairs.sort((a, b) => a[0] - b[0]);

    // Union-Find to connect until all in one circuit
    const uf = new UnionFind(n);
    let numComponents = n;

    for (const [distSq, i, j] of pairs) {
        if (uf.union(i, j)) {  // Actually merged two components
            numComponents--;
            if (numComponents === 1) {
                // This was the last connection - all in one circuit now
                return points[i][0] * points[j][0];  // Product of X coordinates
            }
        }
    }

    return 0;
}

function main() {
    const inputFile = process.argv[2] || join(__dirname, '../input.txt');
    const points = parseInput(inputFile);

    console.log(`Part 1: ${part1(points)}`);
    console.log(`Part 2: ${part2(points)}`);
}

main();
